specs/appwindow-spec.md

```diff
@@ -0,0 +1,93 @@
+Window.AppWindow api
+===
+New api to simplify accessing [appwindow](https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/windowing/windowing-overview)
+functionality through WinUI code
+
+# Background
+Xaml has a [Window](https://docs.microsoft.com/uwp/api/Windows.UI.Xaml.Window) API that
+internally wraps an hwnd. Windows has an [AppWindow](https://docs.microsoft.com/uwp/api/Windows.UI.WindowManagement.AppWindow) class
+that similarly wraps an hwnd in UWP. WinAppSDK has a new [AppWindow](https://docs.microsoft.com/windows/windows-app-sdk/api/winrt/Microsoft.UI.Windowing.AppWindow) which wraps an hwnd and works on Desktop.
+
+You can get an AppWindow from a Xaml Window by calling a COM API to get Xaml and then
```

Comment:
I believe that we need deterministic ordering of subclassing because the “last one wins” creates unexpected ordering that changes when other WinAppSDK components start taking new dependencies.

Let’s say that the app is currently the first one to ask for an AppWindow. No WinAppSDK components ask for an AppWindow today, so it’s never created. The App creates one, so AppWindow gets the “new first” WNDPROC handling (#1 AW). Then, the remaining WNDPROC handlers execute: ContentAppWindowBridge (#2 CAWB) and Xaml’s DesktopWindow (#3 XDW). Each of these WNDPROC handlers exposes some events to the application. So AppWindow (#1 AW)’s events fire first, then ContentAppWindowBridge (#2 CAWB)’s event fire, then Xaml’s DesktopWindow (#3 XDW), and finally DefWindowProc (#4 DWP). (Pretend that all these WinRT objects are synchronously firing events. I’m using it for illustration…):

#1 AW
#2 CAWB
              #2.5 (And currently, Lifted Input internally subclasses as well 😲)
#3 XDW
#4 DWP


Now, let’s say that Xaml’s DesktopWindow starts internally using AppWindow functionality before the ContentAppWindowBridge is created. In this case, ContentAppWindowBridge will create the “new first” WNDPROC when the first DesktopChildSiteBridge is created. In this situation, the ordering will be different:

#1 CAWB
#2 AW
#3 XDW
#4 DWP

Now, let’s say that Xaml completely switches over to AppWindow for the creation of its HWND, and not just subclassing after creation:

#1 CAWB
#2 AW
???
#3 DWP

Now, let’s say that Zach’s DisplayInformation is also subclassing the HWND to fire notifications when the display has changed. And AppWindow’s DisplayAreaWatcher. And anything else… If only the app is creating these various objects, they won’t change. However, as soon as other WinAppSDK components start (correctly) taking dependencies on one another, the ordering can change because the underlying subclass ordering has changed.

Working with Francis, our proposal has been to allow deterministic ordering between WindowFeatures, the new AppWindow subclassing functionality. This would allow WindowFeatures to come and go at runtime, but their ordering is determined by feature ordering, not by last-come subclassing. So even if CAWB was created before or after AppWindow, it wouldn’t matter. We could determine that only WindowFeatures subclasses the HWND, and it ensures that all WindowFeatures will execute in a well-defined ordering.

Reply:
Discussion: Order of subclassing is important. In this spec we're creating the AppWindow during Window construction, in order to make it deterministic. We're not saying here what the exact ordering is amongst multiple parts of the system that are subclassing.

Recommend: Make clear the deterministic ordering not just of the subclassing but of more app-visible APIs, like ordering of related events between AppWindow and Window.